### PR TITLE
CSPL-1924: add azure blob in the app frame work

### PR DESCRIPTION
**Pull request by kumarajeet**
Changes for CSPL-1924: Extend the current app framework for Azure Blob.

Summary of work done:

While looking into the S3 and AzureBlob specs in terms of inputs and outputs, both storage types are pretty much similar. I have added a comparison for S3, Azureblob and Google Storage in this sheet: [data providers similarities](https://docs.google.com/spreadsheets/d/1LMK6L4KH3nih8VSwTfPBFvc6BKYNSqw9hihy3QNLHFA/edit?usp=sharing)

Earlier we were registering `AWSS3Client` and `MinioS3Client` as valid `S3Client` providers, and with this change, we are also registering a new client called `AzureBlobClient`.  Further, as S3Client specs are similar to what Azure and Google storage have, I have generalized the `S3Client` name to `RemoteDataClient` so the interface becomes more generic covering AWSS3, MinioS3, AzureBlob, and in the future, we can plug in `GoogleStorage` client within it if we implement the google storage also.

In summary following implementation 

`S3Client` <------ `AWSS3Client`, `MinioS3Client`

has changed to:

`RemoteDataClient` <------ `AWSS3Client`, `MinioS3Client`, `AzureBlobClient`

changes

You will see lots of places where S3 Client phrase is changed to Remote Data Client. However, some places may need further revision. 

Also, I have added AzureBlobClient and its respective Mock client but they are currently placeholders to be implemented in the next sprint. 

Also added an azure_blob_rest_service that will be covered in the next sprint (list/download) for hashing out the implementation of azure rest APIs usage.

I have made sure all the existing unit tests are passing.

### CODE REVIEW

File: pkg/splunk/client/azureblobclient.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	Timeout      int //timeout in seconds
+}
+
+// Config contains configuration for httpclient
+type Config struct {
+	MaxRetries int         // max number of retries for the rest calls
+	HTTPClient http.Client // the http client to use for calling the rest end points
+}
+
+// ListBlobsOptions contains options for listign apps
```
Comment (akondur) [inline pkg/splunk/client/azureblobclient.go:43]: Nit: listing vs listign

File: pkg/splunk/client/azure_blob_rest_service.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+// 	  Timeout - time in ms that a http rest call should wait before terminating
+//    MaxRetries - number of retries for the rest call if it fails for first time.
+//    http transport related parameters
+// Output:
+//  DownloadBlobOutput
+//    Response Body
+//    Response Code
+//    Total Number of bytes returned
+
+type azureRestService interface {
```
Comment (akondur) [inline pkg/splunk/client/azure_blob_rest_service.go:35]: Is this a mimic of the DownloadApp and ListApps in azureblobclient.go?

Comment (kumarajeet) [reply pkg/splunk/client/azure_blob_rest_service.go:35]: AzureBlobClient (and also similar clients like AWSS3Client, and MinioClient) provide implementation to the methods GetAppsList and DownloadApp 
```
// RemoteDataClient is an interface to provide
// listing and downloading of apps from remote data storage
//
type RemoteDataClient interface {

	// Get the list of Apps
	GetAppsList(context.Context) (RemoteDataListResponse, error)

	// Download a given app as per the inputs provided in the `RemoteDataClientRequest`
	DownloadApp(context.Context, RemoteDataDownloadRequest) (bool /* return pass/fail */, error)
}
'''

These clients will eventually use a third-party data provider. For AzureBlobClient, we are going to use our own small package (currently there is only one file azure_blob_rest_service.go in this package) to list and download apps vis Azure Rest Calls. The `DownloadApp` and `ListApps` methods will be part of this package.

File: pkg/splunk/client/azureblobclient.go (hunk 1 above)
Comment (kumarajeet) [reply pkg/splunk/client/azureblobclient.go]: Thanks Arjun. Corrected it.

File: pkg/splunk/client/azure_blob_rest_service.go (hunk 2 above)
Comment (akondur) [reply pkg/splunk/client/azure_blob_rest_service.go:35]: @kumarajeet Do we need a separate package for it i.e we could have only helper functions for http operations in azure_blob_rest_service.go while keeping the list and download Apis in the azurebloblclient.go?

File: pkg/splunk/client/azure_blob_rest_service.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
+package client
```
Comment (sgontla) [inline pkg/splunk/client/azure_blob_rest_service.go:1]: nit: copyright?

File: pkg/splunk/client/azure_blob_rest_service.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+// ListApps ... TODO: placeholder for next sprint
+func ListApps(ctx context.Context, bucketName string, opts ListBlobsOptions) {
+}
+
+// DownloadApp ... TODO: placeholder for next sprint
+func DownloadApp(ctx context.Context, bucketName string, remoteFile string, localFile string, opts DownloadOptions) {
+
+}
+
+// func (azureblobClient *AzureBlobClient) DownloadBlob(ctx context.Context, downloadRequest RemoteDataDownloadRequest) (bool, error) {
```
Comment (sgontla) [inline pkg/splunk/client/azure_blob_rest_service.go:55]: nit: remove it for now?

File: pkg/splunk/client/azureblobclient.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+	defer file.Close()
+
+	azureBlobClient := client.Client
+
+	//TODO : revise in next sprint . these are just place holder to more
+	// specific data structure so we can pass download options to rest calls
+	downloadOptions := make(map[string]string)
+
+	downloadOptions["maxChunkSize"] = "5gb"
+	downloadOptions["timeout"] = "5sec"
```
Comment (sgontla) [inline pkg/splunk/client/azureblobclient.go:221]: Are we considering to make them configurable? similarly for listing optoins as well.

File: pkg/splunk/client/remotedataclient.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+type RemoteObject struct {
+	Etag         *string
+	Key          *string
+	LastModified *time.Time
+	Size         *int64
+	StorageClass *string
+}
+
+// RemoteDataListRequest struct contains inputs specifying storage account
+//
```
Comment (sgontla) [inline pkg/splunk/client/remotedataclient.go:46]: nit: remove

File: pkg/splunk/client/remotedataclient.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+// RemoteDataListResponse struct contains list of RemoteObject objects
+type RemoteDataListResponse struct {
+	Objects []*RemoteObject
+}
+
+// RemoteDataDownloadRequest struct specifies the remote data file path,
+// local file where the downloaded data should be written as well as
+// the authontication data if available
+type RemoteDataDownloadRequest struct {
+	LocalFile  string // file name where the remote data will be written
```
Comment (sgontla) [inline pkg/splunk/client/remotedataclient.go:59]: file name OR path?

File: pkg/splunk/client/remotedataclient.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	RemoteFile string // file name with path relative to the bucket
+	Etag       string // unique tag of the object
+}
+
+// RemoteDataClient is an interface to provide
+// listing and downloading of apps from remote data storage
+//
+type RemoteDataClient interface {
+
+	// Get the list of Apps
```
Comment (sgontla) [inline pkg/splunk/client/remotedataclient.go:69]: nit: I think, generally, we call it as App package, instead of App. Same with other places.

**Review by sgontla — APPROVED** [review]

File: pkg/splunk/client/azure_blob_rest_service.go (hunk 2 above)
Comment (kumarajeet) [reply pkg/splunk/client/azure_blob_rest_service.go:35]: @akondur the simplest solution will be to keep the rest helper functions in the same file.
However, the way currently the Mockxxxxclients are working (some cyclic package inclusions constraints), we may have to keep the azure_blob_rest_service.go in its own package.

**Review by akondur — APPROVED** [review]

File: pkg/splunk/client/azureblobclient.go (hunk 5 above)
Comment (kumarajeet) [reply pkg/splunk/client/azureblobclient.go:221]: I will refine the use of options in the next sprint. My current thinking is that the Azure rest APIS provides options to break a big blob and download data in smaller chunks so it is possible to download apps in chunks. S3 sdk also has similar features, but we have not used them so far. 
Timeout setting is for http request timeouts, and I may just use the default values to start with (same as we did leave them to default in Aws /Minio)

File: pkg/splunk/client/remotedataclient.go (hunk 6 above)
Comment (kumarajeet) [reply pkg/splunk/client/remotedataclient.go]: done

File: pkg/splunk/client/remotedataclient.go (hunk 7 above)
Comment (kumarajeet) [reply pkg/splunk/client/remotedataclient.go]: thanks. corrected the names.

File: pkg/splunk/client/remotedataclient.go (hunk 8 above)
Comment (kumarajeet) [reply pkg/splunk/client/remotedataclient.go]: changed

File: pkg/splunk/client/azure_blob_rest_service.go (hunk 4 above)
Comment (kumarajeet) [reply pkg/splunk/client/azure_blob_rest_service.go:70]: I plan to have azure_blob_rest_service.go  refactored in the next sprint (while doing list story). So this whole file may be removed .

File: pkg/splunk/test/azureblobclient.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	enterpriseApi "github.com/splunk/splunk-operator/api/v3"
+)
+
+// MockAzureBlobClient is used to store all the objects for an app source
+type MockAzureBlobClient struct {
+	Objects []*MockS3Object
```
Comment (gaurav-splunk) [inline pkg/splunk/test/azureblobclient.go:28]: is this MockS3Object or MockBlobObject?

File: pkg/splunk/enterprise/afwscheduler.go [hunk 10]
```diff
@@ -429,7 +429,7 @@ func (downloadWorker *PipelineWorker) createDownloadDirOnOperator(ctx context.Co
 }
 
 // download API will do the actual work of downloading apps from remote storage
-func (downloadWorker *PipelineWorker) download(ctx context.Context, pplnPhase *PipelinePhase, s3ClientMgr S3ClientManager, localPath string, downloadWorkersRunPool chan struct{}) {
+func (downloadWorker *PipelineWorker) download(ctx context.Context, pplnPhase *PipelinePhase, s3ClientMgr RemoteDataClientManager, localPath string, downloadWorkersRunPool chan struct{}) {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/afwscheduler.go:432]: s3ClientMgr -> remoteDataClientMgr ?

File: pkg/splunk/enterprise/indexercluster_test.go [hunk 11]
```diff
@@ -1284,9 +1284,9 @@ func TestIndexerClusterWithReadyState(t *testing.T) {
 	err = os.MkdirAll(newpath, os.ModePerm)
 
 	// adding getapplist to fix test case
-	GetAppsList = func(ctx context.Context, s3ClientMgr S3ClientManager) (splclient.S3Response, error) {
-		s3Response := splclient.S3Response{}
-		return s3Response, nil
+	GetAppsList = func(ctx context.Context, s3ClientMgr RemoteDataClientManager) (splclient.RemoteDataListResponse, error) {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/indexercluster_test.go:1287]: s/s3ClientMgr/remoteDataClientMgr

File: pkg/splunk/enterprise/licensemaster_test.go [hunk 12]
```diff
@@ -368,20 +368,20 @@ func TestLicensemanagerGetAppsListForAWSS3ClientShouldNotFail(t *testing.T) {
...
 			},
-			getS3Client: func(ctx context.Context, client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkRef *enterpriseApi.AppFrameworkSpec, vol *enterpriseApi.VolumeSpec, location string, fn splclient.GetInitFunc) (splclient.SplunkS3Client, error) {
+			getRemoteDataClient: func(ctx context.Context, client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkRef *enterpriseApi.AppFrameworkSpec, vol *enterpriseApi.VolumeSpec, location string, fn splclient.GetInitFunc) (splclient.SplunkRemoteDataClient, error) {
 				c, err := GetRemoteStorageClient(ctx, client, cr, appFrameworkRef, vol, location, fn)
 				return c, err
 			},
 		}
 
-		s3Response, err := s3ClientMgr.GetAppsList(ctx)
+		RemoteDataListResponse, err := s3ClientMgr.GetAppsList(ctx)
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/licensemaster_test.go:377]: ditto, please change the s3ClientMgr everywhere.

File: pkg/splunk/client/util.go [hunk 13]
```diff
@@ -78,14 +78,38 @@ func NewMockMinioS3Client(ctx context.Context, bucketName string, accessKeyID st
...
+	var s3SplunkClient SplunkAzureBlobClient
+	var err error
+
+	cl := fn(ctx, endpoint, storageAccountName, secretAccessKey)
+	if cl == nil {
+		err = fmt.Errorf("failed to create an Azure blob client")
+		return nil, err
+	}
+
+	s3SplunkClient = cl.(SplunkAzureBlobClient)
```
Comment (gaurav-splunk) [inline pkg/splunk/client/util.go:93]: s/s3SplunkClient/azureSplunkClient ?

File: pkg/splunk/client/awss3client.go [hunk 14]
```diff
@@ -35,7 +34,7 @@ import (
 )
 
 // blank assignment to verify that AWSS3Client implements S3Client
```
[the remark this answers is not in the document]
Comment (kumarajeet) [reply pkg/splunk/client/awss3client.go]: done

File: pkg/splunk/enterprise/util.go [hunk 15]
```diff
@@ -719,67 +719,73 @@ func DeleteOwnerReferencesForS3SecretObjects(ctx context.Context, client splcomm
...
-		return s3Response, err
+		return remoteDataListResponse, err
 	}
-	return s3Response, nil
+	return remoteDataListResponse, nil
 }
 
 // DownloadApp downloads the app from remote storage
-func (s3mgr *S3ClientManager) DownloadApp(ctx context.Context, remoteFile string, localFile string, etag string) error {
+func (s3mgr *RemoteDataClientManager) DownloadApp(ctx context.Context, remoteFile string, localFile string, etag string) error {
```
[the remark this answers is not in the document]
Comment (kumarajeet) [reply pkg/splunk/enterprise/util.go:752]: done

File: pkg/splunk/enterprise/util.go [hunk 16]
```diff
@@ -719,67 +719,73 @@ func DeleteOwnerReferencesForS3SecretObjects(ctx context.Context, client splcomm
...
 }
 
-// GetAppListFromS3Bucket gets the list of apps from remote storage.
-func GetAppListFromS3Bucket(ctx context.Context, client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkRef *enterpriseApi.AppFrameworkSpec) (map[string]splclient.S3Response, error) {
+// GetAppListFromRemoteBucket gets the list of apps from remote storage.
+func GetAppListFromRemoteBucket(ctx context.Context, client splcommon.ControllerClient, cr splcommon.MetaObject, appFrameworkRef *enterpriseApi.AppFrameworkSpec) (map[string]splclient.RemoteDataListResponse, error) {
 
 	reqLogger := log.FromContext(ctx)
-	scopedLog := reqLogger.WithName("GetAppListFromS3Bucket").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+	scopedLog := reqLogger.WithName("GetAppListFroGetAppListFromRemoteBucketS3Bucket").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
```
[the remark this answers is not in the document]
Comment (kumarajeet) [reply pkg/splunk/enterprise/util.go]: done

File: pkg/splunk/test/azureblobclient.go (hunk 9 above)
Comment (kumarajeet) [reply pkg/splunk/test/azureblobclient.go:28]: Good point! My plan is revise the test for azureblobclient.go during list api story, So let me take a look in it then if that's ok with you.

File: pkg/splunk/enterprise/licensemaster_test.go (hunk 12 above)
Comment (kumarajeet) [reply pkg/splunk/enterprise/licensemaster_test.go]: done

File: pkg/splunk/client/util.go (hunk 13 above)
Comment (kumarajeet) [reply pkg/splunk/client/util.go]: done

File: pkg/splunk/client/azureblobclient.go [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+	return nil, nil
+}
+
+// DownloadApp to be revsited
+func (client *SplunkAzureBlobClientTemp) DownloadApp(ctx context.Context, bucketName string, remoteFileName string, localFileName string, downloadOpts map[string]string) error {
+	return nil
+}
+
+// AzureBlobClient is a client to implement Azure Blob specific APIs
+type AzureBlobClient struct {
```
Comment (vivekr-splunk) [inline pkg/splunk/client/azureblobclient.go:77]: suggestion, can we separate structures to separate files as models.  like for example azureblobclien_types.go which only contains structures defines.

File: pkg/splunk/client/azureblobclient.go [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+	StorageAccountName string
+	SecretAccessKey    string
+	Prefix             string
+	StartAfter         string
+	Endpoint           string
+	Client             SplunkAzureBlobClient
+}
+
+// NewAzureBlobClient returns an AzureBlob client
+func NewAzureBlobClient(ctx context.Context, bucketName string, storageAccountName string, secretAccessKey string, prefix string, startAfter string, region string, endpoint string, fn GetInitFunc) (RemoteDataClient, error) {
```
Comment (vivekr-splunk) [inline pkg/splunk/client/azureblobclient.go:88]: instead of passing so many arguments should we pass a structure , like `BucketCredentials` or `BucketDetails`

File: pkg/splunk/client/minioclient.go [hunk 19]
```diff
@@ -47,14 +47,14 @@ type MinioClient struct {
 }
 
 // NewMinioClient returns an Minio client
-func NewMinioClient(ctx context.Context, bucketName string, accessKeyID string, secretAccessKey string, prefix string, startAfter string, region string, endpoint string, fn GetInitFunc) (S3Client, error) {
+func NewMinioClient(ctx context.Context, bucketName string, accessKeyID string, secretAccessKey string, prefix string, startAfter string, region string, endpoint string, fn GetInitFunc) (RemoteDataClient, error) {
```
Comment (vivekr-splunk) [inline pkg/splunk/client/minioclient.go:50]: Same here, instead of passing so many arguments, should we consider using structure of `MinioBucketCredentials`

File: pkg/splunk/client/azureblobclient.go [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+	err := json.Unmarshal(listBlobsXML, &(remoteDataListResponse.Objects))
+	if err != nil {
+		return remoteDataListResponse, err
+	}
+
+	return remoteDataListResponse, nil
+}
+
+// DownloadApp downloads an app package from remote storage
+func (client *AzureBlobClient) DownloadApp(ctx context.Context, downloadRequest RemoteDataDownloadRequest) (bool, error) {
```
Comment (vivekr-splunk) [inline pkg/splunk/client/azureblobclient.go:202]: I am unsure why we need 2 return values here. err itself should be suffice right ?

File: pkg/splunk/client/minioclient.go [hunk 21]
```diff
@@ -73,8 +73,8 @@ func NewMinioClient(ctx context.Context, bucketName string, accessKeyID string,
 
 //RegisterMinioClient will add the corresponding function pointer to the map
 func RegisterMinioClient() {
-	wrapperObject := GetS3ClientWrapper{GetS3Client: NewMinioClient, GetInitFunc: InitMinioClientWrapper}
-	S3Clients["minio"] = wrapperObject
+	wrapperObject := GetRemoteDataClientWrapper{GetRemoteDataClient: NewMinioClient, GetInitFunc: InitMinioClientWrapper}
+	RemoteDataClientsMap["minio"] = wrapperObject
```
Comment (vivekr-splunk) [inline pkg/splunk/client/minioclient.go:77]: I have my doubts in using map here. I do not think we ever plan to support multiple cloud services at the same time ? . are we over designing here ?

File: pkg/splunk/client/azureblobclient.go (hunk 17 above)
Comment (kumarajeet) [reply pkg/splunk/client/azureblobclient.go:77]: Good point Vivek. I can check this in upcoming sprints. I just followed the other clients existing layout (like AWSS3Client).

File: pkg/splunk/client/azureblobclient.go (hunk 18 above)
Comment (kumarajeet) [reply pkg/splunk/client/azureblobclient.go:88]: Like the idea. Let me revisit this in one of the upcoming sprints if that's ok with you.

File: pkg/splunk/client/minioclient.go (hunk 19 above)
Comment (kumarajeet) [reply pkg/splunk/client/minioclient.go:50]: Like the idea. Let me revisit this in one of the upcoming sprints if that's ok with you.

File: pkg/splunk/client/azureblobclient.go (hunk 20 above)
Comment (kumarajeet) [reply pkg/splunk/client/azureblobclient.go:202]: Good point. I used the same approach as other clients like AWSS3Client :) Happy to look into it in one of the future sprints. There are some places the callers are using the bool return value also which seems to be redundant. Legacy code.

File: pkg/splunk/client/minioclient.go (hunk 21 above)
Comment (kumarajeet) [reply pkg/splunk/client/minioclient.go:77]: Per my discussion with some of the folks (@gaurav-splunk / @sgontla ), plan is to keep support for multiple cloud services. it looks a little vague but if one cloud provider (e.g aks cluster) has access to another cloud provider (like AWS S3) then our design should work. More to be covered in testing.

Comment (gaurav-splunk) [reply pkg/splunk/client/minioclient.go:77]: We do support multiple cloud providers for different volumes for the same CRD, hence this.

**Review by gaurav-splunk — APPROVED** [review]
